theano/compile/function.py

```diff
@@ -185,6 +185,22 @@ def opt_log1p(node):
 
 
     """
+    if isinstance(outputs, dict):
+        output_items = outputs.items()
+
+        output_items_sorted = sorted(output_items)
```

Comment:
In the latest version of the code I restricted it to strings.  Python3 does indeed raise an exception when sorting a list containing both strings and integers.  

The only non-string type that I could see some value in using for keys is a tuple.  For example: 

{("h", "dropout") : h_dropout, ("h", "no dropout"): h_no_dropout}

but I don't think that this is a big deal.